### DISCUSSION TITLE
Error using oracledb_async driver on Windows 10 x64 (TypeError: Expected bytes, got bytearray)

### DISCUSSION THREAD

**Discussion by Impro02** [sqlalchemy/sqlalchemy#10954] [Usage Questions]
Considering these examples, using dboracle_async driver and an async session on Windows 10 x64, I got the following error when trying to execute a statement. 

I dot not have this issue on linux.

The error is triggered with main_oracledb_async & main_sqla_async, so the error must be on dboracle part.

```py
import asyncio
from sqlalchemy import create_engine, text
from sqlalchemy.ext.asyncio import async_sessionmaker, create_async_engine
from sqlalchemy.orm import sessionmaker
import oracledb

user = "usr"
password = "pw"
dsn = "dsn"


async def main_oracledb():
    connection = oracledb.connect(
        user=user,
        password=password,
        dsn=dsn,
    )

    cursor = connection.cursor()
    cursor.execute("SELECT table_name FROM all_tables")
    while True:
        row = cursor.fetchone()
        if row is None:
            break
        print(row)


async def main_oracledb_async():
    async with oracledb.connect_async(
        user=user, password=password, dsn=dsn
    ) as connection:
        res = await connection.fetchall("SELECT table_name FROM all_tables")
        print(res)


async def main_sqla_async():
    connections_string = f"oracle+oracledb_async://{user}:{password}@{dsn}"

    engine = create_async_engine(
        connections_string,
        echo=False,
    )
    session_factory = async_sessionmaker(
        bind=engine,
        autoflush=False,
        expire_on_commit=False,
    )

    async with session_factory() as session:
        result = await session.execute(text("SELECT table_name FROM all_tables"))

        scalar = result.scalars()

        print(scalar)


async def main_sqla():
    connections_string = f"oracle+oracledb://{user}:{password}@{dsn}"

    engine = create_engine(
        connections_string,
        echo=False,
    )
    session_factory = sessionmaker(
        bind=engine,
        autoflush=False,
        expire_on_commit=False,
    )

    with session_factory() as session:
        result = session.execute(text("SELECT table_name FROM all_tables"))

        scalar = result.scalars()

        print(scalar)

# asyncio.run(main_oracledb())
# asyncio.run(main_oracledb_async())
# asyncio.run(main_sqla_async())
# asyncio.run(main_sqla())
```

```
ERROR:asyncio:Exception in callback _ProactorReadPipeTransport._loop_reading(<_OverlappedF...shed result=8>)
handle: <Handle _ProactorReadPipeTransport._loop_reading(<_OverlappedF...shed result=8>)>
Traceback (most recent call last):
  File "C:\Users\Public\AppData\Local\miniconda3\envs\alphaz-next\Lib\asyncio\events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "C:\Users\Public\AppData\Local\miniconda3\envs\alphaz-next\Lib\asyncio\proactor_events.py", line 324, in _loop_reading
    self._data_received(data, length)
  File "C:\Users\Public\AppData\Local\miniconda3\envs\alphaz-next\Lib\asyncio\proactor_events.py", line 274, in _data_received
    self._protocol.data_received(data)
  File "src\\oracledb\\impl/thin/protocol.pyx", line 895, in oracledb.thin_impl.BaseAsyncProtocol.data_received
TypeError: Expected bytes, got bytearray
```

**Comment by CaselIT** [comment]
Hi,

Without more information there isn't much that can be suggested. An example of model and query would be required.

Does this error happens also when trying a similar query using the oracle driver directly?

**Reply by Impro02** [reply]
I updated the post to add examples, the error is triggered while using oracledb directly.

**Reply by CaselIT** [reply]
thanks! mike already pinged the driver developers, since this looks like something that comes from the driver. Thanks for updating the example

**Comment by zzzeek** [comment]
cc @anthony-tuininga , as you were looking for things to help with ! :)

**Reply by anthony-tuininga** [reply]
Looks to me like the `data_received` method is not receiving `bytes` but `bytearray`. The asyncio [documentation](https://docs.python.org/3/library/asyncio-protocol.html#asyncio.Protocol.data_received) suggests that it should be `bytes`. The [code](https://github.com/python/cpython/blob/main/Lib/asyncio/proactor_events.py#L191) is showing the use of a `bytearray` however. I'm not sure whether the documentation should be modified to indicate that `bytes` and `bytearray` are possible or whether the code should be transforming the `bytearray` into `bytes` before calling the function to conform to the documentation. Any thoughts?

**Reply by anthony-tuininga** [reply]
Looking at the CPython code it looks like a bug in CPython -- the documentation for the `Protocol` class clearly states that the argument to `data_received` is a `bytes object` (https://github.com/python/cpython/blob/main/Lib/asyncio/protocols.py#L97).

**Reply by zzzeek** [reply]
OK a few things, how come the user has a different result under linux than windows?  is this a different python version and the issue was fixed later?  I see for example `C:\Users\Public\AppData\Local\miniconda3\envs\alphaz-next\Lib\asyncio\proactor_events.py`, i.e. "conda" where we see all kinds of versioning / build problems all the time

**Reply by CaselIT** [reply]
The proactor event loop, that I'm assuming it implemented in `proactor_events`, is a windows only thing.
Linux uses another implementation. 

> Looking at the CPython code it looks like a bug in CPython

Would be interesting to understand if it's always been like that or it was introduced in a particular python version.

**Reply by Impro02** [reply]
There is that issue that was closed in cpython: https://github.com/python/cpython/issues/99941

I tried on python 3.10.x and 3.11.x, both have the same behavior.

**Reply by CaselIT** [reply]
from the linked issue at some point they mention that "bytes object" means any bytes but then the PR changes things to bytes. Maybe that spot was missed in the PR that fixed that issue?

**Reply by CaselIT** [reply]
> Maybe that spot was missed in the PR that fixed that issue?

or it's on 3.12+ only that's fixed? Locally when implementing the async oracledb driver I tested only on 3.12 on windows, so that may be a possibility

@Impro02 could you try to check if 3.12 works for you?

**Reply by Impro02** [reply]
I confirm it works on python 3.12+
